clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -446,14 +453,15 @@ boolean joinGroupIfNeeded(final Timer timer) {
                     resetJoinGroupFuture();
                     needsJoinPrepare = true;
                 } else {
-                    log.info("Generation data was cleared by heartbeat thread. Initiating rejoin.");
+                    log.info("Generation data was cleared by heartbeat thread to {} and state is now {} before " +
+                         "the rebalance callback is triggered, marking this rebalance as failed and retry",
+                         generation, state);
                     resetStateAndRejoin();
                     resetJoinGroupFuture();
-                    return false;
                 }
             } else {
                 final RuntimeException exception = future.exception();
-                log.info("Join group failed with {}", exception.toString());
+                log.info("Rebalance failed with {}", exception.toString());
```

Comment:
Can we still specify that we failed during the JoinGroup? eg `Rebalance failed on JoinGroup with {}` or something

Reply:
The reason I changed it is exactly that it may not always due to join-group :) If sync-group failed, this could also be triggered.

Reply:
Different question: can we report the exception as the "cause", rather than just getting the toString of it?
```suggestion
                log.info("Rebalance failed.", exception);
```